opm/parser/eclipse/EclipseState/NumericalAquifer.hpp

```diff
@@ -0,0 +1,111 @@
+/*
+  Copyright (C) 2020 Equinor ASA
+  Copyright (C) 2020 SINTEF Digital
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  ?You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef OPM_NUMERICAL_AQUIFER_HPP
+#define OPM_NUMERICAL_AQUIFER_HPP
+
+#include <vector>
+#include <unordered_map>
+#include <set>
+#include <array>
+
+namespace Opm {
+    class Deck;
+    class DeckRecord;
+    class NumericalAquiferConnections;
+    struct NumAquiferCon;
+    class EclipseGrid;
+    class FieldPropsManager;
+    class NNC;
+    namespace Fieldprops {
+        class TranCalculator;
+    };
+
+    struct NumericalAquiferCell {
+        NumericalAquiferCell(const DeckRecord&, const EclipseGrid&, const FieldPropsManager&);
+        size_t aquifer_id; // aquifer id
+        size_t I, J, K; // indices for the grid block
+        double area; // cross-sectional area
+        double length;
+        double porosity;
+        double permeability;
+        double depth; // by default the grid block depth will be used
+        double init_pressure; // by default, the grid pressure from equilibration will be used
+        int pvttable; // by default, the block PVTNUM
+        int sattable; // saturation table number, by default, the block value
+        double pore_volume; // pore volume
+        double transmissibility;
+        size_t global_index;
+    };
+
+    class SingleNumericalAquifer {
+    public:
+        explicit SingleNumericalAquifer(const size_t aqu_id);
+        void addAquiferCell(const NumericalAquiferCell& aqu_cell);
+        void addAquiferConnection(const NumAquiferCon& aqu_con);
+        void updateCellProps(const EclipseGrid& grid,
```

Comment:
The way I understand it the `updateCellProps` method should be called by the simulator to update the `pore_volume, satnum, pvtnum` and `cell_depth` arrays? Allowing the opm-common aquifer implementation to overwrite the core data in this way is in my opinion quite an invasion of privacy. I think you should rather export a vector of small structs and then let the simulator tinker with it's own data; i.e. something like:
```
// opm-common:

struct AquiferCell {
      std::size_t global_index;
      double pore_volume; 
      int satnum;
      int pvtnum;
      double cell_depth;
}
 
std::vector<AquiferCell> exportCellProps(const EclipseGrid& grid) {
}

// simulator
auto aquifer_cells = aquifer_object.exportCellProps(grid);

for (const auto& aq_cell : aquifer_cells) {
     ....
}
```

Reply:
I understand `updateCellProps` is invasive.  The main way I hope this functionality works is that we fix all (as much as possible) the information when we parsing it, then at the simulator side, we do not need to do much or nothing. 

From the parser to the simulator, we have a lot of good implementation and design there, hopefully the functionality can work  seamless with all these good implementation and design. 

If we check the `aquifer_cells` whenever we touch the aquifer cells, there might be a lot of places we need to do it and it will be invasive in another manner and might not be easy to get things correct. 

There is some challenges here is that in the simulator code might use different ways to obtain information. For example, `cell_depth` is not used in the simulator code, which makes that we have to check the `aquifer_cells` to get the depth information.

Reply:
OK - we will take our time and find a good solution together. Thank you for opening the PR early.